app/logic/updateTrackHours.py

```diff
@@ -0,0 +1,75 @@
+# from app.controllers.admin import admin_bp
+from app.models.eventParticipant import EventParticipant
+from datetime import datetime
+from app.models.user import User
+
+
+def getEventLengthInHours(startTime, endTime, eventDate):
+    """
+    Converts the event length hours into decimal
+    parameters: startTime- start time event (type: time)
+                endTime- end time event (type: time)
+                eventDate- date of the event (type: datetime)
+    """
+    #can only subtract datetime objects, not time objects. So convert time into datetime
+    eventLength = datetime.combine(eventDate, endTime) - datetime.combine(eventDate, startTime)
+    (h, m, s) = str(eventLength).split(':')
+    eventLengthInHours = int(h) + round(int(m) / 60, 2)
+
+    return eventLengthInHours
+
+
+def updateTrackHours(participantData):
+    """
+    updates the events participant table with data from the track hours webpage
+
+    param: participantData- a dictionary that contains data from every row of the page along with the associated username.
+    """
+
+    for user in range(1, len(participantData)):
+        if f'username{user}' in participantData:
+            username = participantData[f'username{user}']
+            if (User.select().where(User.username == username)).exists():
+
+                try:
+                    if participantData['checkbox_'+ username] == "on": #if the user is marked as present
+                        (EventParticipant.update({EventParticipant.hoursEarned: float(participantData['inputHours_'+ username])}).where(EventParticipant.event == participantData['event'],
+                                                 EventParticipant.user == participantData[f'username{user}'])).execute()
+
+                        (EventParticipant.update({EventParticipant.attended: True}).where(EventParticipant.event == participantData['event'],
+                                                 EventParticipant.user == participantData[f'username{user}'])).execute()
+                except:   #if there is no checkbox for user then they are not present for the event.
+                    (EventParticipant.update({EventParticipant.attended: False}).where(EventParticipant.event == participantData['event'],
+                        EventParticipant.user == participantData[f'username{user}'])).execute()
+
+                    (EventParticipant.update({EventParticipant.hoursEarned: 0}).where(EventParticipant.event == participantData['event'],
+                                             EventParticipant.user == participantData[f'username{user}'])).execute()
+            else:
+                return "Volunteer does not exist."
+        else:
+            break
+
+    return "Volunteer table successfully updated!"

```

Comment:
I'm not a big fan of functions returning text messages about success or failure. Those messages should belong to the controller, not the application functions